project_setup/scripts/iam-to-github

```diff
@@ -0,0 +1,247 @@
+            secret: str = resource["values"]["secret"]
+            user: str = resource["values"]["user"]
+            yield user, key_id, secret
+    return
+
+
+def creds_from_terraform() -> Generator[Tuple[str, str, str], None, None]:
+    """Retrieve IAM credentials from terraform state.
+
+    Yields (user, key_id, secret) when found.
+    """
+    user: str
+    key_id: str
+    secret: str
+    c = subprocess.run(["terraform", "show", "--json"], stdout=subprocess.PIPE)  # nosec
+    # Normally we'd check the process return code here.  But terraform is perfectly
+    # happy to return zero even if there was no state files.
+    j: Dict = json.loads(c.stdout)
+
+    if not j.get("values"):
+        return
+
+    for child_module in j["values"]["root_module"]["child_modules"]:
+        for user, key_id, secret in creds_from_child(child_module):
+            yield user, key_id, secret
+    return
+
+
+def encrypt(public_key: str, secret_value: str) -> str:
+    """Encrypt a Unicode string using the public key."""
+    public_key = public.PublicKey(public_key.encode("utf-8"), encoding.Base64Encoder())
+    sealed_box = public.SealedBox(public_key)
+    encrypted = sealed_box.encrypt(secret_value.encode("utf-8"))
+    return b64encode(encrypted).decode("utf-8")
+
+
+def get_public_key(session: requests.Session, repo_name) -> Dict[str, str]:
+    """Fetch the public key for a repository."""
+    logging.info(f"Requesting public key for repository {repo_name}")
+    response = session.get(
+        f"https://api.github.com/repos/{repo_name}/actions/secrets/public-key"
+    )
+    response.raise_for_status()
+    return response.json()
+
+
+def set_secret(
+    session: requests.Session,
+    repo_name: str,
+    secret_name: str,
+    secret_value: str,
+    public_key: Dict[str, str],
+) -> None:
+    """Create a secret in a repository."""
+    logging.info(f"Creating secret {secret_name}")
+    encrypted_secret_value = encrypt(public_key["key"], secret_value)
+    response = session.put(
+        f"https://api.github.com/repos/{repo_name}/actions/secrets/{secret_name}",
+        json={
+            "encrypted_value": encrypted_secret_value,
+            "key_id": public_key["key_id"],
+        },
+    )
+    response.raise_for_status()
+
+
+def get_repo_name() -> str:
+    """Get the repository name using git."""
+    logging.debug(f"Trying to determine GitHub repository name using git.")
+    c = subprocess.run(  # nosec
+        ["git", "remote", "get-url", "origin"],
+        stdout=subprocess.PIPE,
+        stderr=subprocess.PIPE,
+    )
+    if c.returncode != 0:
+        logging.critical("Could not determine GitHub repository name.")
+        raise Exception(c.stderr)
+    match = GIT_URL_RE.match(c.stdout.decode())
+    repo_name = match.groups()[0]  # type: ignore
+    return repo_name
+
+
+def main() -> int:
+    """Set up logging and call the requested commands."""
+    args: Dict[str, str] = docopt.docopt(__doc__, version="1.0.0")
```

Comment:
Should this be `Dict[str, Any]` to be consistent with assignment to `validated_args` on line 166? I think this would also be consistent with the fact that arguments returned by `docopt` are `Dict[str, Any]` in general.

Reply:
Not yet... docopt acts like `sys.argv` at this point and is not coercing types.  After we validate with `Schema` the signature of the arguments will be cast to their appropriate types.

https://github.com/cisagov/development-guide/blob/18cda9463081e3019013f895a529d5560f8f6517/project_setup/scripts/iam-to-github#L166

Reply:
Ah, ok. That makes sense.